### PR TITLE
Preconditions: add generic `check()` that allows to supply a custom exception

### PR DESCRIPTION
**Pull request by schildbach**
The idea is to use this in #2821 to throw an `ArithmeticException` rather than a `IllegalArgumentException`.

### CODE REVIEW

**Comment by schildbach** [comment]
This PR is now kind of without direct use, as #2821 will not be merged.

However it might be useful on its own. I recently added tests to this PR.

File: core/src/main/java/org/bitcoinj/base/internal/Preconditions.java [hunk 1]
```diff
@@ -25,8 +25,7 @@ public class Preconditions {
      * @throws IllegalArgumentException if {@code expression} is false
      */
     public static void checkArgument(boolean expression) {
-        if (!expression)
-            throw new IllegalArgumentException();
+        check(expression, () -> new IllegalArgumentException());
```
Comment (msgilligan) [inline core/src/main/java/org/bitcoinj/base/internal/Preconditions.java:28]: I think you can use `IllegalArgumentException::new` here.

Comment (msgilligan) [inline core/src/main/java/org/bitcoinj/base/internal/Preconditions.java:28]: I think you can use `IllegalArgumentException::new` here.

**Review by msgilligan — CHANGES_REQUESTED** [review]
Looks good. One minor suggestion.

File: core/src/main/java/org/bitcoinj/base/internal/Preconditions.java (hunk 1 above)
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/base/internal/Preconditions.java:28]: Thanks, added.

Comment (schildbach) [reply core/src/main/java/org/bitcoinj/base/internal/Preconditions.java:28]: Added.

**Comment by schildbach** [comment]
Merged.

**Pull request closed without merging**